packages/tasit-action/src/TasitAction.js

```diff
@@ -20,76 +20,85 @@ class Utils {
-      if (event.eventName === eventName && event.listener === listener) {
-        this.#emitter.removeListener(
-          event.wrappedEventName,
-          event.wrappedListener
-        );
-        return false;
-      }
-      return true;
-    });
+  _toEthersEventName = eventName => {
+    if (eventName === "confirmation") return "block";
+    return eventName;
   };
 
-  off = (eventName, listener) => {
-    this.removeListener(eventName, listener);
-  };
+  off = eventName => {
+    const event = this.#events.get(eventName);
+
+    if (!event) return;
 
-  listenerCount = eventName => {
-    return this.#events.filter(event => {
-      return event.eventName === eventName;
-    }).length;
+    if (eventName !== "error") {
+      const { listener } = event;
+
+      // Note: ethers.removeAllListeners(name) not working!
+      this.#ethersEventEmitter.removeListener(
+        this._toEthersEventName(eventName),
+        listener
+      );
+    }
+    this.#events.delete(eventName);
   };
 
-  removeAllListeners = () => {
-    this.#events.forEach(event => {
-      this.removeListener(event.eventName, event.listener);
+  unsubscribe = () => {
+    this.#events.forEach((event, eventName) => {
+      this.off(eventName);
     });
   };
 
-  unsubscribe = () => {
-    this.removeAllListeners();
+  eventNames = () => {
+    return Array.from(this.#events.keys());
   };
 
   // TODO: Make protected
-  emitErrorEvent = error => {
-    this.#events.forEach(event => {
-      if (event.eventName === "error") {
-        event.listener(error);
-      }
-    });
+  _emitErrorEvent = (error, eventName) => {
+    const errorEvent = this.#events.get("error");
+    if (!errorEvent) {
+      // Note: Throw error?
+      console.warn(`Error emission without listener: ${error}`);
+      return;
+    }
+
+    const message = { error, eventName };
+    errorEvent.listener(message);
   };
 
   // TODO: Make protected
-  addErrorListener = listener => {
-    const eventName = "error";
-    this.#events.push({
-      eventName,
-      wrappedEventName: eventName,
+  _addErrorListener = listener => {
+    this.#events.set("error", {
       listener,
-      wrappedListener: listener,
     });
   };
 
   // TODO: Make protected
-  addListener = (eventName, wrappedEventName, listener, wrappedListener) => {
-    this.#emitter.on(wrappedEventName, wrappedListener);
+  _addEventListener = (eventName, listener) => {
+    if (eventName === "error")
+      throw new Error(
+        `Use _addErrorListener function to subscribe an error event.`
+      );
+
+    if (this.eventNames().includes(eventName))
+      throw new Error(`Event '${eventName}' already registred.`);
 
-    this.#events.push({
-      eventName,
-      wrappedEventName,
+    this.#events.set(eventName, {
       listener,
-      wrappedListener,
     });
+
+    this.#ethersEventEmitter.on(this._toEthersEventName(eventName), listener);
+  };
+
+  // For testing purposes
```

Comment:
Do you think we want to create a `tech debt` issue for extracting functions used just for testing from the main source code files? Or do you think it's a fine pattern to leave this function here?

I'm just curious in an ideal world - either way it's a great compromise for now

Reply:
Ah, this is the place we're discussing where functions that are just for testing purposes go

Reply:
I'll create an issue, can we keep those functions for now? At least while we are on unstable versions?

Reply:
https://github.com/tasitlabs/TasitSDK/issues/126

Reply:
Yep! Sounds good. Thanks. Resolving this.